packages/core/src/validators.ts

```diff
@@ -1033,12 +1050,21 @@ export class IsNullOrUndefinedValidator extends Validator {
 
 export class NotEmptyValidator extends Validator {
   validatePath(value: any, path: Path, ctx: ValidationContext): PromiseLike<ValidationResult> {
-    return !isNullOrUndefined(value) && isNumber(value.length) && value.length > 0
-      ? ctx.successPromise(value)
-      : ctx.failurePromise(defaultViolations.notEmpty(path), value);
+    if (isNullOrUndefined(value) || (isObject(value) && !Object.keys(value).length) || (hasNumericLength(value) && !value.length)) {
+      return ctx.failurePromise(defaultViolations.notEmpty(path, value), value);
+    }
+    return ctx.successPromise(value);
   }
 }
 
+function isObject(value: any) {
+  return typeof value === 'object' && value !== null && !Array.isArray(value);
+}
+
+function hasNumericLength(value: any) {
+  return (Array.isArray(value) || typeof value === 'string') && typeof value.length === 'number';
```

Comment:
This method name is a bit misleading. Essentially it's a check that the value is an array or string. As both those allowed types have `length` the `typeof value.length === 'number'` check is redundant. Also it does not count for `new String("")` (see `isString` function). Maybe it would be more clear to state that this validator supports strings, arrays and objects, that they are not null, undefined or empty? The old implementation, i.e. the existence of `length` property, is unclear in case of an object: is `{ length: 0 }` empty or not?